db/model/sample.js

```diff
@@ -168,29 +171,42 @@ module.exports = function sample(seq, dataTypes) {
        * doesn't invoke hooks, and we rely on hooks to publish the change.
        *
        * @param {Sample} toUpsert - The sample to upsert
+       * @param {Object} req - Request object from the api layer with
+       * the authorization
+       * @param {String} - userName - User name using which the upesert
+       * operation is made. This is used to check if the user has access to
+       * perform the upsert operation
        * @param {Boolean} isBulk - true when used with bulk upsert, false
        *  otherwise
        * @returns {Promise} which resolves to the newly-created or -updated
        *  sample record, or rejects if a ResourceNotFoundError was encountered
        *  looking up the associated subject and aspect, or if an error was
        *  encountered while performing the sample upsert operation itself.
        */
-      upsertByName(toUpsert, isBulk) {
+      upsertByName(toUpsert, userName, isBulk) {
```

Comment:
This is the only place we're doing the write permission check in the db layer--everywhere else is in the api layer. At the very least, we should have a comment explaining why this needs to be here and not in API layer.

That being said, it would be nice to put the write-permission checking into custom middleware that gets executed for patch/put/delete operations on the designated routes *after* the passport check but *before* the requests are even routed to the API controller. Let's discuss.